### DISCUSSION TITLE
Support for AI Language Servers (e.g., Copilots) in LSP Server

### DISCUSSION THREAD

**Discussion by bajrangCoder** [mkslanc/ace-linters#111] [General]
This project is fantastic! It is compatible with a ton of coding language servers like rust-analyzer, gopls, and more.

I'm curious, though. Does it also support AI through LSP? For example, GitHub Copilot and Codium work on editors like Vim and Neovim using language servers. I gave it a shot with GitHub Copilot, but it didn't seem to work. Are there any plans to integrate this feature?

In this era of AI, it's really important. It could be incredibly helpful for developers to build an LSP server that works universally across major editors.

Some references:

- https://github.com/Exafunction/codeium.nvim
- https://github.com/Exafunction/codeium-chrome
- https://github.com/github/copilot.vim
- https://github.com/TerminalFi/LSP-copilot?tab=readme-ov-file

**Comment by mkslanc** [comment]
Hmm, I'll investigate, it shouldn't be anyhow different from other language servers

**Reply by bajrangCoder** [reply]
Yes , they are similar

**Comment by mkslanc** [comment]
I couldn't find any standalone language server, everything is just hard-coded plugin/extension for different IDE (which are simply wrapper on API). Do you know any?

**Reply by bajrangCoder** [reply]
Everything are just compiled binaries, there is no any actual code available for the ai LSP 😔

Btw check out this : https://stackoverflow.com/questions/76741410/how-to-invoke-github-copilot-programmatically

And clients of other editors are compatible such as code mirror, monaco, etc, Why not this one?

Recently Zed code editor also added support for these GitHub copilot through binary LSP server , here is [source code](https://github.com/zed-industries/zed)

There is one more project similar to this one which is completely* open source, [here](https://github.com/continuedev/continue)

**Comment by bajrangCoder** [comment]
...

**Reply by mkslanc** [reply]
I'll consider adding support, but not right now. Currently I have my hands full with other stuff..

**Reply by bajrangCoder** [reply]
Ok No problem 👍 

But also checkout this one : https://github.com/SilasMarvin/lsp-ai

I had tried it but whenever I connect LSP-ai through websocket to ace it just fails with some unclear errors

**Reply by mkslanc** [reply]
lsp-ai is implemented by following lsp, so it should work through webSocket's bridge as any other server

**Reply by bajrangCoder** [reply]
But that doesn't work , whenever it connected the the editor through websocket then it just breaks.

I just used your ace linters demo (svelte) and modified initialisation option and websocket stuff and used your this [repo](https://github.com/mkslanc/websockets-ls) as server and added lsp-ai command in stdio mode

**Reply by mkslanc** [reply]
There were few small bugs in client implementation, new version should support this.
- [Installation page](https://github.com/mkslanc/ace-linters/wiki/LSP%E2%80%90AI-for-Ace%E2%80%90linters)
- [Full example](https://github.com/mkslanc/ace-linters/blob/main/packages/demo/websockets-lsp-ai/client.ts)

I played and it seems a bit bumpy at the moment - server is not supporting abort request (it's on roadmap though), so you need carefully configure `liveAutocompletionDelay` not to eat all your credits, if you are using OpenAI or Atrophic. 
I also tried local llm, but seems my laptop couldn't handle load normally or I haven't configured model normally. It took ~1 minute for response for simple request.

But other than that - it's working, so try to enjoy :)

**Reply by bajrangCoder** [reply]
I have tried it with groq endpoint and the lsp-ai is providing completion (it can be seen in websocket message tab of inspect tool) but that completion isn't rendered on ace .

I am missing something? 🤔 
I have used LSP ai demo from this repo

**Reply by mkslanc** [reply]
There are annoying bug with showing inner completion first time, try ctrl +space first to manually trigger completion

**Reply by bajrangCoder** [reply]
Hmm, it's working 👍, (is there no any way to fix that issue)

**Reply by mkslanc** [reply]
Better to solve it on ace side, the inline autocompletion code is quite confusing

**Reply by bajrangCoder** [reply]
Hey 👋 
@mkslanc GitHub recently published their [language server ](https://www.npmjs.com/package/@github/copilot-language-server) to npm and also there are information regarding how to integrate with other editors. So I think now it will work with ace editor(through ace-linters). I haven't tried but I'm sure that it'll now work.

**Comment by mkslanc** [comment]
It took some time, but copilot with inline autocomplete is up (https://mkslanc.github.io/ace-linters/copilot.html)

**Reply by bajrangCoder** [reply]
this is awesome :)
